src/GraphQL/Validation/MatchingNodeVisitor.cs

```diff
@@ -43,4 +43,48 @@ void INodeVisitor.Leave(INode node, ValidationContext context)
             }
         }
     }
+
+    /// <summary>
+    /// A node listener which runs configured delegates only when the node entered/left matches the specified node type.
+    /// </summary>
+    /// <typeparam name="TNode">A specified AST node type.</typeparam>
+    /// <typeparam name="TState">Type of the provided state.</typeparam>
+    public class MatchingNodeVisitor<TNode, TState> : INodeVisitor
```

Comment:
internal?

Reply:
Wouldn't this be used for any third-party validation rule?

Reply:
Oh this is a new class.....let me look at it...

Reply:
So it is only used by GetVariables?  I guess 50/50 to me.  I'd probably leave it public I suppose but whatever you think.